File: day01.py
import re

f = open('inputs/input01.txt', 'r')
lines = f.read().splitlines()                       # read lines
lines = [re.sub(r'\D', '', line) for line in lines] # strip non-numbers
lines = [line[0] + line[-1] for line in lines]      # get first and last
lines = [int(line) for line in lines]               # convert to integer
print("Part A:", sum(lines))

def replace_digit_names(lines):
  """
  Isn't this a beautifully elegant function?
  """
  lines = [line.replace('one',   'one1one')     for line in lines]
  lines = [line.replace('two',   'two2two')     for line in lines]
  lines = [line.replace('three', 'three3three') for line in lines]
  lines = [line.replace('four',  'four4four')   for line in lines]
  lines = [line.replace('five',  'five5five')   for line in lines]
  lines = [line.replace('six',   'six6six')     for line in lines]
  lines = [line.replace('seven', 'seven7seven') for line in lines]
  lines = [line.replace('eight', 'eight8eight') for line in lines]
  lines = [line.replace('nine',  'nine9nine')   for line in lines]
  return lines


f.seek(0) # go back to first line
lines = f.read().splitlines()                       # read lines
lines = replace_digit_names(lines)                  # replace digit names
lines = [re.sub(r'\D', '', line) for line in lines] # strip non-numbers
lines = [line[0] + line[-1] for line in lines]      # get first and last
lines = [int(line) for line in lines]               # convert to integer

print("Part B:", sum(lines))

File: day01/day01.rb
# frozen_string_literal: true

input = File.readlines('./input01.txt')

# Part A
lines = input
        .map { _1.scan(/\d/) }
        .map { (_1[0] + _1[-1]).to_i }

puts "Part A: #{lines.inject(&:+)}"

# Part B

def replace_digit_names(str)
  str = str.gsub('one',   'one1one')
  str = str.gsub('two',   'two2two')
  str = str.gsub('three', 'three3three')
  str = str.gsub('four',  'four4four')
  str = str.gsub('five',  'five5five')
  str = str.gsub('six',   'six6six')
  str = str.gsub('seven', 'seven7seven')
  str = str.gsub('eight', 'eight8eight')
  str.gsub('nine', 'nine9nine')
end

lines = input
        .map { replace_digit_names(_1) }
        .map { _1.gsub(/\D/, '')}
        .map { _1.scan(/\d/) }
        .map { (_1[0] + _1[-1]).to_i }

puts "Part B: #{lines.inject(&:+)}"
